## ****A/B Testing Analysis for Targeted Coupon Campaign****

In [2]:
import pandas as pd
import numpy as np

In [3]:
file_path = r'C:\Users\19802\Documents\学习\Cornell_AB Testing_25Fall\jcb540s_tool-coupon.xlsx'
sheet_name = 'coupon'
coupon = pd.read_excel(r'C:\Users\19802\Documents\学习\Cornell_AB Testing_25Fall\jcb540s_tool-coupon.xlsx', sheet_name)

In [4]:
print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  
0       0.0   138         3  110.60       0   54  
1       0.0   328         1   94.60       1   51  
2       0.0   349         3  175.00       1   47  
3       0.0   122         5  314.23       1   45  
4       0.0     9         4   48.91       1   44  


In [5]:
#Create a new column 'coupon' to indicate better coupon= 2, base coupon =1, no coupon =0
coupon['coupon'] = coupon['group'].apply(
    lambda x: 1 if x in ['T1_L', 'T1_H'] 
    else 2 if x in ['T2_H'] 
    else 0
)

print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  coupon  
0       0.0   138         3  110.60       0   54       0  
1       0.0   328         1   94.60       1   51       2  
2       0.0   349         3  175.00       1   47       2  
3       0.0   122         5  314.23       1   45       1  
4       0.0     9         4   48.91       1   44       1  


In [6]:
#Transform to dummy variable
#Create a new column 'high' to indicate high-value segment
coupon['high'] = coupon['group'].apply(
    lambda x: 1 if x in ['T1_H', 'T2_H', 'CG_H']
    else 0
)

print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  coupon  high  
0       0.0   138         3  110.60       0   54       0     1  
1       0.0   328         1   94.60       1   51       2     1  
2       0.0   349         3  175.00       1   47       2     1  
3       0.0   122         5  314.23       1   45       1     1  
4       0.0     9         4   48.91       1   44       1     0  


In [7]:
#Transform to dummy variable
#Create new columns 'base' and 'better' as dummy variables
coupon['base'] = coupon['group'].apply(
    lambda x: 1 if x in ['T1_L', 'T1_H']
    else 0
)

coupon['better'] = coupon['group'].apply(
    lambda x: 1 if x in ['T2_H']
    else 0
)

print(coupon.head())

   customerid group  visit  pageviews  duration  redeem  purchase  amount  \
0   100003171  CG_H      0          0         0       0         0     0.0   
1   100005146  T2_H      0          0         0       0         0     0.0   
2   100010245  T2_H      0          0         0       0         0     0.0   
3   100016186  T1_H      0          0         0       0         0     0.0   
4   100016657  T1_L      0          0         0       0         0     0.0   

   discount  last  numpurch  totdol  gender  age  coupon  high  base  better  
0       0.0   138         3  110.60       0   54       0     1     0       0  
1       0.0   328         1   94.60       1   51       2     1     0       1  
2       0.0   349         3  175.00       1   47       2     1     0       1  
3       0.0   122         5  314.23       1   45       1     1     1       0  
4       0.0     9         4   48.91       1   44       1     0     1       0  


In [8]:
#Create a low_value seg as df
coupon['high'].dtype

low_value_segment = coupon[coupon['high'] == 0]

print(low_value_segment)

       customerid group  visit  pageviews  duration  redeem  purchase  amount  \
4       100016657  T1_L      0          0         0       0         0     0.0   
5       100021299  T1_L      1          2       175       0         0     0.0   
8       100024643  T1_L      0          0         0       0         0     0.0   
9       100028202  CG_L      0          0         0       0         0     0.0   
11      100028818  T1_L      0          0         0       0         0     0.0   
...           ...   ...    ...        ...       ...     ...       ...     ...   
12949   125469864  CG_L      1          4        70       0         0     0.0   
12951   125474475  CG_L      0          0         0       0         0     0.0   
12955   125538236  T1_L      0          0         0       0         0     0.0   
12956   125563783  CG_L      0          0         0       0         0     0.0   
12958   125585750  T1_L      0          0         0       0         0     0.0   

       discount  last  nump

In [9]:
coupon['totdol'].describe()

count    12959.000000
mean        97.677053
std         75.443554
min         17.600000
25%         33.300000
50%         77.900000
75%        136.825000
max        381.320000
Name: totdol, dtype: float64

In [10]:
#Get summary statistics for the column 'totdol' group by base coupon or no coupon for low_value segment
low_value_segment.groupby('base')['totdol'].describe()

count       mean        std   min     25%   50%     75%   max
base                                                                
0     1998.0  39.877735  17.660587  18.6  25.600  35.0  51.150  87.2
1     4787.0  40.301959  18.166353  17.6  25.375  34.6  51.875  87.4

In [11]:
high_value_segment = coupon[coupon['high'] == 1]

## ****Low Value Segment Analysis****

### ***Verify Random Assignment for test vs control group for all the pre-test variables***

*Base vs Control*

Other ways to check random assignment:
- Check the method used: is a random generator/algorithm used?
- Visual - similar shape and spreads

In [12]:
#low-value base
low_value_segment[low_value_segment['base'] == 1]['group'].value_counts()

group
T1_L    4787
Name: count, dtype: int64

In [13]:
low_base = low_value_segment[low_value_segment['base'] == 1]

In [14]:
#low-value control
low_value_segment[low_value_segment['base'] == 0]['group'].value_counts()

group
CG_L    1998
Name: count, dtype: int64

In [15]:
low_control = low_value_segment[low_value_segment['base'] == 0]

In [16]:
#Verify Random Assignment for test vs control group for all the pre-test variables 

from scipy import stats

#1.totdol - total spend in past 12 months
#Null hypothesis: there is no relationship between group assignment (test/control) and spending patterns.
t_stat, p_value = stats.ttest_ind(low_control['totdol'], low_base['totdol'], equal_var=True)  # set equal_var=False for Welch’s t-test
print("p-value for totdol:", p_value)

#2.last - number of days since last purchase
t_stat, p_value = stats.ttest_ind(low_control['last'], low_base['last'], equal_var=True)  # set equal_var=False for Welch’s t-test
print("p-value for last:", p_value)

#3. numpurch - number of purchases in past 12 months
t_stat, p_value = stats.ttest_ind(low_control['numpurch'], low_base['numpurch'], equal_var = True)
print("p-value for numpurch: ", p_value)

#4. gender
t_stat, p_value = stats.ttest_ind(low_control['gender'], low_base['gender'], equal_var = True)
print("p-value for gender: ", p_value)

#5. age
t_stat, p_value = stats.ttest_ind(low_control['age'], low_base['age'], equal_var = True)
print("p-value for age: ", p_value)

p-value for totdol: 0.3767616186103123
p-value for last: 0.15877966942962546
p-value for numpurch:  0.19509760653807492
p-value for gender:  0.5469569068782603
p-value for age:  0.04000608354890501


***In conclusion, P values for totdol, last, numpurch, and gender are all higher than 0.5 means that the null hypothesis cannot be rejected, which means randomization was achieved in the low-value segment between test and control groups. However, in terms of age, p value being 0.04, which is lower than 0.05. Null hypothesis is being rejected, which means randomization was NOT achieved in the low-value segment in terms of age. It means there is not enough evidence to support that the test and control groups are sufficiently randomized by age. Any conclusion of coupon use based on coupon use for low-value segment could be off.***

### ***Analyze Purchase Behavior***

*Base vs Control*

In [17]:
low_value_segment['group'].value_counts()

group
T1_L    4787
CG_L    1998
Name: count, dtype: int64

In [18]:
low_value_segment.groupby('group')['purchase'].mean()*100

group
CG_L    1.001001
T1_L    2.465009
Name: purchase, dtype: float64

In [19]:
summary1 = (
    low_value_segment.groupby('group').agg(
        customers= ('purchase', 'count'),  # number of observations
        purchase_rate= ('purchase', lambda x: x.mean() * 100),   # % of 1s if binary 
        redeem_rate = ('redeem', lambda x: x.mean() * 100),
        ARPC = ('amount', 'mean') # Average Revenue per Customer   
    ))

print(summary1)

       customers  purchase_rate  redeem_rate      ARPC
group                                                 
CG_L        1998       1.001001     0.000000  0.254585
T1_L        4787       2.465009     0.480468  0.713378


In [20]:
#low_value buyer df
low_value_buyer = low_value_segment[low_value_segment['purchase'] == 1]

In [21]:
summary2 = (
    low_value_buyer.groupby('group')['amount'].mean().reset_index(name='ARPB')
)

print(summary2)

  group       ARPB
0  CG_L  25.433000
1  T1_L  28.940169


In [22]:
summary = summary1.merge(summary2, on='group', how='left')

print(summary.round(2))

  group  customers  purchase_rate  redeem_rate  ARPC   ARPB
0  CG_L       1998           1.00         0.00  0.25  25.43
1  T1_L       4787           2.47         0.48  0.71  28.94


In [23]:
#Calculate p value for purchase_rate
#Ho: Base coupon doesn't affect the purchase rate for the low-value segment
#Ha: Base coupon significantly affect the purchase rate for the low-value segment

p_purchase= stats.ttest_ind(low_control['purchase'], low_base['purchase'], equal_var=True).pvalue # set equal_var=False for Welch’s t-test
print("p-value for purchase:", p_purchase)

p-value for purchase: 9.801375180873792e-05


In [24]:
summary['p_purchase_rate'] = summary['group'].map({
    'CG_L': None,
    'T1_L': p_purchase
})

print(summary.round(4))

  group  customers  purchase_rate  redeem_rate    ARPC     ARPB  \
0  CG_L       1998          1.001       0.0000  0.2546  25.4330   
1  T1_L       4787          2.465       0.4805  0.7134  28.9402   

   p_purchase_rate  
0              NaN  
1           0.0001  


The p-value for purchase rate is 0.001, less than 0.05, so we can make the conclusion that base coupon significantly affect the purchase rate for the low-value segment from 1% to 2.47%.

In [25]:
#Calculate p value for amount (ARPC)
p_ARPC= stats.ttest_ind(low_control['amount'], low_base['amount'], equal_var=True).pvalue  # set equal_var=False for Welch’s t-test
print("p-value for ARPC:", p_ARPC)

p-value for ARPC: 0.000835227839360256


In [26]:
summary['p_ARPC'] = summary['group'].map({
    'CG_L': None,
    'T1_L': p_ARPC
})

print(summary.round(4))

  group  customers  purchase_rate  redeem_rate    ARPC     ARPB  \
0  CG_L       1998          1.001       0.0000  0.2546  25.4330   
1  T1_L       4787          2.465       0.4805  0.7134  28.9402   

   p_purchase_rate  p_ARPC  
0              NaN     NaN  
1           0.0001  0.0008  


The p-value for APRC is 0.008, less than 0.05, so we can make the conclusion that the base coupon significantly increases the revenue for the low-value segment from 0.26 to 0.71.

In [27]:
#Calculate p value for ARPB

low_buyer_control = low_value_segment[(low_value_segment['purchase'] == 1) & (low_value_segment['base'] == 0)]
low_buyer_base = low_value_segment[(low_value_segment['purchase'] == 1) & (low_value_segment['base'] == 1)]
p_ARPB = stats.ttest_ind(low_buyer_control['amount'], low_buyer_base['amount'], equal_var = True).pvalue

print("p value for ARPB:", p_ARPB)

p value for ARPB: 0.526535606699631


In [28]:
summary['p_ARPB'] = summary['group'].map({
    'CG_L': None,
    'T1_L': p_ARPB
})

print(summary.round(4))

  group  customers  purchase_rate  redeem_rate    ARPC     ARPB  \
0  CG_L       1998          1.001       0.0000  0.2546  25.4330   
1  T1_L       4787          2.465       0.4805  0.7134  28.9402   

   p_purchase_rate  p_ARPC  p_ARPB  
0              NaN     NaN     NaN  
1           0.0001  0.0008  0.5265  


The p-value for ARPB is 0.52, higher than 0.05, we cannot reject null hypothesis, meaning that there is no statistically significant impact on average revenue per buyer for low-value segment when they receive base coupon.

***Conclusion: For low-value segment, the base coupon is effective in increasing purchase rate and purchase amount, but not effective in increasing the purchase amount among the buyers. In other words, the base coupon increases the revenue among the low-value segment because of the increase in the pruchase rate, but doesn't make buyers spend more.***

### ***Analyze the Connversion Funnel***

*Base vs Control*

In [103]:
summary_low = pd.DataFrame()

In [104]:
summary_low = (
    low_value_segment.groupby('group').agg(
        customers= ('visit', 'count'),  # number of observations
        visit_rate= ('visit', lambda x: x.mean() * 100)   # % of 1s if binary 
    ).reset_index())

print(summary_low)

  group  customers  visit_rate
0  CG_L       1998   11.211211
1  T1_L       4787   35.115939


In [105]:
#Calculate p-value for visit rate
p_visit = stats.ttest_ind(low_control['visit'], low_base['visit'], equal_var = True).pvalue
print(p_visit)

2.273346804665243e-91


In [106]:
summary_low['p_visit'] = summary_low['group'].map({
    'CG_L': None,
    'T1_L': p_visit
})

print(summary_low.round(4))

  group  customers  visit_rate  p_visit
0  CG_L       1998     11.2112      NaN
1  T1_L       4787     35.1159      0.0


The p-value for visit rate is close to 0, which mean we can reject the null hypothesis. The base coupon effectively increase website visits 3 times in the low-value segment.

In [107]:
#The other variables, like page view, duration and purchase are conditional on visit = 1

low_visitor = low_value_segment[low_value_segment['visit'] == 1]
low_visitor_control = low_value_segment[(low_value_segment['visit'] == 1) & (low_value_segment['base'] == 0)]
low_visitor_base = low_value_segment[(low_value_segment['visit'] == 1) & (low_value_segment['base'] == 1)]

In [108]:
#Avg Pageviews
pageviews = (
    low_visitor.groupby('group')['pageviews'].mean().round(2)
)

summary_low = summary_low.merge(pageviews, on='group', how='left')

print(summary_low)

  group  customers  visit_rate       p_visit  pageviews
0  CG_L       1998   11.211211           NaN      12.54
1  T1_L       4787   35.115939  2.273347e-91       9.04


In [109]:
#p-value for page views
p_pageviews = stats.ttest_ind(low_visitor_control['pageviews'], low_visitor_base['pageviews'], equal_var = True).pvalue

print(p_pageviews)

7.998338362842542e-05


In [110]:
summary_low['p_pageviews'] = summary_low['group'].map({
    'CG_L': None,
    'T1_L': p_pageviews
})

print(summary_low.round(4))

  group  customers  visit_rate  p_visit  pageviews  p_pageviews
0  CG_L       1998     11.2112      NaN      12.54          NaN
1  T1_L       4787     35.1159      0.0       9.04       0.0001


The p-value for pageviews is 0.0001, lower than 0.05, which mean we can reject the null hypothesis. The base coupon effectively DECREASES website pageviews in the low-value segment, even as it creases websites visits overall.

In [111]:
#Duration
duration = (
    low_visitor.groupby('group')['duration'].mean().round(2)/3600
)

summary_low = summary_low.merge(duration, on='group', how='left')

print(summary_low)

  group  customers  visit_rate       p_visit  pageviews  p_pageviews  duration
0  CG_L       1998   11.211211           NaN      12.54          NaN   0.22100
1  T1_L       4787   35.115939  2.273347e-91       9.04      0.00008   0.18125


In [112]:
p_duration = stats.ttest_ind(low_visitor_control['duration'], low_visitor_base['duration'], equal_var = True).pvalue

print(p_duration)

0.12724022871726506


In [113]:
summary_low['p_duration'] = summary_low['group'].map({
    'CG_L': None,
    'T1_L': p_duration
})

print(summary_low.round(4))

  group  customers  visit_rate  p_visit  pageviews  p_pageviews  duration  \
0  CG_L       1998     11.2112      NaN      12.54          NaN    0.2210   
1  T1_L       4787     35.1159      0.0       9.04       0.0001    0.1812   

   p_duration  
0         NaN  
1      0.1272  


The p-value for duration is 0.1272, higher than 0.05, we cannot reject the null. Therefore, the difference in duration between control and base groups are not statistically significant.

In [116]:
#Purchase %
purchase_rate = (
low_visitor.groupby('group')['purchase'].mean()*100
)

summary_low = summary_low.merge(purchase_rate, on = 'group', how = 'left')

print(summary_low)

  group  customers  visit_rate       p_visit  pageviews  p_pageviews  \
0  CG_L       1998   11.211211           NaN      12.54          NaN   
1  T1_L       4787   35.115939  2.273347e-91       9.04      0.00008   

   duration  p_duration  purchase  
0   0.22100         NaN  8.928571  
1   0.18125     0.12724  7.019631  


In [118]:
p_purchase_rate = stats.ttest_ind(low_visitor_control['purchase'], low_visitor_base['purchase'], equal_var = True).pvalue

print(p_purchase_rate)

0.30075184169374425


In [119]:
summary_low['p_purchase_rate'] = summary_low['group'].map({
    'CG_L': None,
    'T1_L': p_purchase_rate
})

print(summary_low.round(4))

  group  customers  visit_rate  p_visit  pageviews  p_pageviews  duration  \
0  CG_L       1998     11.2112      NaN      12.54          NaN    0.2210   
1  T1_L       4787     35.1159      0.0       9.04       0.0001    0.1812   

   p_duration  purchase  p_purchase_rate  
0         NaN    8.9286              NaN  
1      0.1272    7.0196           0.3008  


The p-value for purchase rate is 0.3, higher than 0.05. The base coupon is not effective in impacting purchase behavior among website visitors in the low-value segment.

***Conclusion: For low-value segment, the base coupon is effective in increasing website visit rate and decreasing the page views, but not effective in increasing the website duration and purchase rate among website visitors.***

### ***Linear Regression***

After conducting  A/B testing, the purpose of creating a linear regression model is to gain deeper insights and quantify the relationship between variables.

1- Adjust for Covariates (control variables): a linear model allows you to control for these covariates (eg. age, location, gender), isolating the true effect of the treatment.

2- Estimate the Effect Size: Gives you an estimated lift (β coefficient) for the test group.

3- Predictive Power: Linear models can be used to predict future outcomes based on the treatment and other variables.

In [18]:
import statsmodels.api as sm

In [19]:
#For low-value segment, is there a relationsihp between receiving the base coupon and purchase amount?
#base vs amount

X = low_value_segment[['base']]
y = low_value_segment['amount']

# Add constant for intercept
X = sm.add_constant(X)

# Fit OLS model
linear_low= sm.OLS(y, X).fit()

# Print summary
print(linear_low.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     11.17
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           0.000835
Time:                        06:47:47   Log-Likelihood:                -20752.
No. Observations:                6785   AIC:                         4.151e+04
Df Residuals:                    6783   BIC:                         4.152e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2546      0.115      2.208      0.0

1. Test the overall significance of the regression model compared to a model with no independent varilabes.
   The P-value for the model is 0.000835 --> Reject null, meaning that the regression model could be valuable.
2. Assess how good the regression model is.
   Adjusted R-squared is 0.001 --> The regression model can explain about 0.1% of the information in the dependent variable.
3. Test the significance of each individual coefficient
   the coef is 0.4588, and p value is 0.001 --> reject null, meaning the base coupon is a significant predictor of purchase amount resulting in a lift of 0.46, which matches the difference between low_base and low_control in ARPC.
4. Make predictions.
   

***Conclusion: the base coupon increases the revenue by 0.46 on average, compared to control.***

In [20]:
# Make predictions (like R's predict(..., type='response'))
amount_hat = linear_low.predict(X)

# Summary statistics (like R's summary(amount_hat))
print(pd.Series(amount_hat).describe())

count    6785.000000
mean        0.578276
std         0.209136
min         0.254585
25%         0.254585
50%         0.713378
75%         0.713378
max         0.713378
dtype: float64


**Question: for low value segment, is there still a relationship between receiving the base coupon and purchase amount once you account for other customer characteristics?**


In [27]:
X = low_value_segment[['base', 'last', 'numpurch','totdol', 'gender', 'age']]
y = low_value_segment['amount']

X = sm.add_constant(X)

linear2_low = sm.OLS(y, X).fit()

print(linear2_low.summary())

                            OLS Regression Results                            
Dep. Variable:                 amount   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.004
Method:                 Least Squares   F-statistic:                     5.154
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           2.69e-05
Time:                        07:02:30   Log-Likelihood:                -20742.
No. Observations:                6785   AIC:                         4.150e+04
Df Residuals:                    6778   BIC:                         4.155e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.4304      0.403     -1.069      0.2

1. Test the overall significance of the regression model compared to a model with no independent varilabes. The P-value for the model is 2.69e-05 --> Reject null, meaning that the regression model is significant.
2. Assess how good the regression model is. Adjusted R-squared is 0.004 --> The regression model can explain about 0.4% of the information in the dependent variable.
Compared with the simple linear regression of 0.001 as Adjusted R-sqaured, this multiple regression model can explain the dependent variable ***better***.
3. Test the significance of each individual coefficient --> base and last are the only two independent variables who have less than 0.05 p-value, meaning the effect of the base coupon and number of days since last purchase for low value segment are significant. I am ignoring age since age is not randomized based my previous analysis.
The base coupon lifts the revenue by 0.44, which is similar to the coefficient of 0.46 in the simple linear regression model, indicating the ***robustness and independence*** of the results.
4. Make predictions

In [29]:
# Make predictions (like R's predict(..., type='response'))
amount_hat2_low = linear2_low.predict(X)

# Summary statistics (like R's summary(amount_hat))
print(pd.Series(amount_hat2_low).describe())

count    6785.000000
mean        0.578276
std         0.347591
min        -0.485545
25%         0.348793
50%         0.581417
75%         0.813294
max         1.967852
dtype: float64
